Refactor Python profiling installation and configuration process (#9651)
DataDog/documentation: content/en/tracing/profiler/getting_started.md

|      Feature         | Supported Python versions          |
|----------------------|------------------------------------|
| Wall time profiling  | Python >= 2.7                      |
| CPU time profiling   | Python >= 2.7 on POSIX platforms   |
| Exception profiling  | Python >= 3.7 on POSIX platforms   |
| Lock profiling       | Python >= 2.7                      |
| Memory profiling     | Python >= 3.5                      |

**Installation**

Install `ddtrace`, which provides both tracing and profiling functionalities:

```shell
pip install ddtrace
```

**Note**: Profiling requires the `ddtrace` library version 0.40+.

If you're using a platform where `ddtrace` binary distribution is not
available, you will need to install a development environment.

For example, on Alpine Linux, this can be done with:
```shell
apk install gcc musl-dev linux-headers
```

**Usage**

To automatically profile your code, set the `DD_PROFILING_ENABLED` environment
variable to `true` when you use `ddtrace-run`:

    DD_PROFILING_ENABLED=true \
    DD_ENV=prod \
    DD_SERVICE=my-web-app \
    DD_VERSION=1.0.3 \
    dd-trace-run python app.py

> It is strongly recommended that you add tags like `service` or `version`, as they

Reviewer: I wonder if folks will think they need to do something in addition to the env vars above. Maybe something like

> In addition to the env, service, and version shown above, you can add other tags like `DD_TAGS=canary:true,shard:3` which will provide the ability to slice and dice...

Or maybe we just mention DD_TAGS and leave details for below. I'm not sure, just throwing something out to make it concrete. Non-blocking.

{{< /programming-lang >}}
{{< programming-lang lang="python" >}}

### Requirements

> The Datadog Profiler requires Python 2.7+ and agent to version [7.20.2][1]+ or

Reviewer 2: ```suggestion
The Datadog Profiler requires Python 2.7+ and Agent version [7.20.2][1]+ or
```

[6.20.2][1]+.

> Those profiling features are available depending on your Python version:

Reviewer 2: ```suggestion
The following profiling features are available depending on your Python version:
```

> |      Feature         | Supported Python Versions          |

Reviewer 2: ```suggestion
|      Feature         | Supported Python versions          |
```

|----------------------|------------------------------------|
| Wall time profiling  | Python >= 2.7                      |
| CPU time profiling   | Python >= 2.7 on POSIX platforms   |
| Exception profiling  | Python >= 3.7 on POSIX platforms   |
| Lock profiling       | Python >= 2.7                      |
| Memory profiling     | Python >= 3.5                      |

### Installation

> Install `ddtrace` which provides both tracing and profiling functionalities:

Reviewer 2: ```suggestion
Install `ddtrace`, which provides both tracing and profiling functionalities:
```

```shell
pip install ddtrace
```

> **Note**: Profiling requires the `ddtrace` library versions 0.40+.

Reviewer 2: ```suggestion
**Note**: Profiling requires the `ddtrace` library version 0.40+.
```

If you're using a platform where `ddtrace` binary distribution is not
available, you will need to install a development environment.

For example, on Alpine Linux, this can be done with:
```shell
apk install gcc musl-dev linux-headers
```

### Usage

> To automatically profile your code, set `DD_PROFILING_ENABLED` environment

Reviewer 2: ```suggestion
To automatically profile your code, set the `DD_PROFILING_ENABLED` environment
```

variable to `true` when you use `ddtrace-run`:

    DD_PROFILING_ENABLED=true ddtrace-run python app.py

> It is strongly recommended to add tags like `service` or `version` as it

Reviewer 2: ```suggestion
It is strongly recommended that you add tags like `service` or `version`, as they
```

provide the ability to slice and dice your profiles across these dimensions,
enhancing your overall product experience. See [Configuration][#configuration] below.

After a couple of minutes, visualize your profiles on the [Datadog APM > Profiler page][2].

If you want to manually control the lifecycle of the profiler, use the
`ddtrace.profiling.profiler.Profiler` object:

```python
from ddtrace.profiling import Profiler

> prof = Profiler()

Reviewer: And we could do the equivalent here, with comments showing what env var is used if not specified, like
```
prof = Profiler(
    env="prod",  # if not specified, uses env var DD_ENV
...
```

prof.start()

# At shutdown
prof.stop()
```

<div class="alert alert-warning">
The profiler has been designed to be turned on continuously. Only use the
`start` and `stop` methods to start and stop the profiler at the beginning and

> end of your program, not to selectively profile parts of your program.

Reviewer: If we do the above, we can probably ditch this. The less stuff people have to think about when onboarding, the better!

```python
# For ddtrace-run users, call this in your child process
ddtrace.profiling.auto.start_profiler()

# Alternatively, for manual instrumentation,
# create a new profiler in your child process:
from ddtrace.profiling import Profiler

prof = Profiler()
prof.start()
```

**Configuration**

You can configure the profiler using the following environment variables:

| Environment variable                             | Type          | Description                                                                                      |
| ------------------------------------------------ | ------------- | ------------------------------------------------------------------------------------------------ |
| `DD_PROFILING_ENABLED`                           | Boolean       | Set to `true` to enable profiler. |
| `DD_SERVICE`                                     | String        | The Datadog [service][3] name.     |
| `DD_ENV`                                         | String        | The Datadog [environment][4] name, for example, `production`. |
| `DD_VERSION`                                     | String        | The version of your application.                             |

> | `DD_TAGS`                                        | String        | Tags to apply to an uploaded profile. Must be a list of `<key>:<value>` separated by commas such as: `layer:api,team:intake`.   |

Reviewer: Work another col for the programmatic equivalent of each?
